lib/devmapper_mock.ml

```diff
@@ -324,68 +197,101 @@ module DeviceSet = struct
-  try
-    ref (DeviceSet.load_file !persistent_fn)
-  with _ ->
-    ref (DeviceSet.make ())
+  let create device targets =
+    DeviceSet.create !devices device targets;
+    DeviceSet.save_file !devices !persistent_fn
 
-let create device targets =
-  DeviceSet.create !devices device targets;
-  DeviceSet.save_file !devices !persistent_fn
+  let remove device =
+    DeviceSet.remove !devices device;
+    DeviceSet.save_file !devices !persistent_fn
 
-let remove device =
-  DeviceSet.remove !devices device;
-  DeviceSet.save_file !devices !persistent_fn
+  let reload device targets =
+    DeviceSet.reload !devices device targets;
+    DeviceSet.save_file !devices !persistent_fn
 
-let reload device targets =
-  DeviceSet.reload !devices device targets;
-  DeviceSet.save_file !devices !persistent_fn
+  let suspend device =
+    DeviceSet.suspend !devices device;
+    DeviceSet.save_file !devices !persistent_fn
 
-let suspend device =
-  DeviceSet.suspend !devices device;
-  DeviceSet.save_file !devices !persistent_fn
+  let resume device =
+    DeviceSet.resume !devices device;
+    DeviceSet.save_file !devices !persistent_fn
 
-let resume device =
-  DeviceSet.resume !devices device;
-  DeviceSet.save_file !devices !persistent_fn
+  let mknod device path mode =
+    DeviceSet.mknod !devices device path mode;
+    DeviceSet.save_file !devices !persistent_fn
 
-let mknod device path mode =
-  DeviceSet.mknod !devices device path mode;
-  DeviceSet.save_file !devices !persistent_fn
+  let stat device =
+    let ret = DeviceSet.stat !devices device in
+    DeviceSet.save_file !devices !persistent_fn;
+    ret
 
-let stat device =
-  let ret = DeviceSet.stat !devices device in
-  DeviceSet.save_file !devices !persistent_fn;
-  ret
+  let ls () =
+    let ret = DeviceSet.ls !devices () in
+    DeviceSet.save_file !devices !persistent_fn;
+    ret
 
-let ls () =
-  let ret = DeviceSet.ls !devices () in
-  DeviceSet.save_file !devices !persistent_fn;
-  ret
+  let clear () =
+    DeviceSet.clear !devices ();
+    DeviceSet.save_file !devices !persistent_fn
 
-let clear () =
-  DeviceSet.clear !devices ();
-  DeviceSet.save_file !devices !persistent_fn
+  let get_persistent () =
+    !persistent_fn
 
-let get_persistent () =
-  !persistent_fn
+  let set_persistent filename =
+    persistent_fn := filename
 
-let set_persistent filename =
-  persistent_fn := filename
+  let save_file path =
+    DeviceSet.save_file !devices path
 
-let save_file path =
-  DeviceSet.save_file !devices path
+  let load_file path =
+    devices := DeviceSet.load_file path
+end
+
+let create = Mock.create
+let remove = Mock.remove
+let reload = Mock.reload
+let suspend = Mock.suspend
+let resume = Mock.resume
+let mknod = Mock.mknod
+let stat = Mock.stat
+let ls = Mock.ls
+let clear = Mock.clear
+let get_persistent = Mock.get_persistent
+let set_persistent = Mock.set_persistent
+let save_file = Mock.save_file
+let load_file = Mock.load_file
+
+module type DM_intf = sig
```

Comment:
Hi,

I think the module type should be exposed as `Devmapper.S.DEVMAPPER`. It
should work if you put the module type into 's.mli' (no s.ml file needed).
Have a look at this example:
https://github.com/mirage/ocaml-cohttp/blob/master/lib/s.mli

I had forgotten about `Location`, `Striped` and `Target`: thanks for
reminding me! I think we could separate them into separate files so they
appear as `Devmapper.Location` `Devmapper.Striped` and `Devmapper.Target`.
So we would have:

- `Devmapper.S.DEVMAPPER`: the module type
- `Devmapper.Location`: the definition of Location
- `Devmapper.Striped`: the definition of Striped
- `Devmapper.Target`: the definition of Target
- `Devmapper.Linux`: the linux implementation which implements
`Devmapper.S.DEVMAPPER` using `libdevicemapper`
- `Devmapper.Mock`: your new mock implementation which also implements
`Devmapper.S.DEVMAPPER`

This will be quite similar to other projects like [mirage/ocaml-cohttp],
[mirage/ocaml-vchan] and [mirage/mirage-block-volume].

Thanks!
Dave

On Fri, May 22, 2015 at 12:36 PM, Kaifeng Zhu <notifications@github.com>
wrote:

> In lib/devmapper_mock.ml
> <https://github.com/xapi-project/camldm/pull/8#discussion_r30887990>:
>
> > +
> > +let create = Mock.create
> > +let remove = Mock.remove
> > +let reload = Mock.reload
> > +let suspend = Mock.suspend
> > +let resume = Mock.resume
> > +let mknod = Mock.mknod
> > +let stat = Mock.stat
> > +let ls = Mock.ls
> > +let clear = Mock.clear
> > +let get_persistent = Mock.get_persistent
> > +let set_persistent = Mock.set_persistent
> > +let save_file = Mock.save_file
> > +let load_file = Mock.load_file
> > +
> > +module type DM_intf = sig
>
> @djs55 <https://github.com/djs55>, thanks for comments. To enable the
> clients to choose which implementations in run time, should the module type
> S.DEVMAPPER be exposed too? Where should the common module type be placed
> in common OCaml convention?
>
> Also, as submodules (e.g., Location, Striped and Target) exist within
> Devmapper.Linux, I think they are common to Devmapper.Linux and
> Devmapper.Mock. Where should they go then?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xapi-project/camldm/pull/8/files#r30887990>.
>



-- 
Dave Scott